File: backend/app/core/age.py
"""The one age-threshold calculation the platform makes.

Whether someone counts as an adult decides which identity document they hold
(CNIC vs B-Form) and, later, which onboarding paperwork applies to them (bank
details vs a mobile wallet). Every caller derives it from `date_of_birth`
rather than trusting a client-supplied flag, so it lives here once instead of
being re-derived slightly differently in each service that needs it.
"""

from datetime import date


def is_adult(born: date, *, today: date | None = None) -> bool:
    """Whether `born` is 18 or older as of `today` (defaults to the real date)."""
    today = today or date.today()
    age = today.year - born.year - ((today.month, today.day) < (born.month, born.day))
    return age >= 18

File: frontend/src/lib/age.ts
/**
 * Whether someone is 18 or older on a given day, calendar-correct rather than
 * `days / 365.25` (which drifts a day either side of a birthday).
 *
 * The one age-threshold calculation the app makes — decides the CNIC/B-Form
 * requirement at registration and, later, the Bank/Easypaisa split in
 * onboarding. One definition rather than one per call site, so a future
 * change to the rule cannot update three places and miss a fourth. Mirrors
 * backend/app/core/age.py::is_adult; both must change together.
 *
 * Takes `unknown` because every caller reads this straight out of a form
 * value or an API response, neither of which TypeScript can promise is
 * actually a string.
 */
export function isAdult(dateOfBirth: unknown, today: Date = new Date()): boolean {
  if (typeof dateOfBirth !== "string" || !dateOfBirth) return false
  const dob = new Date(dateOfBirth)
  if (Number.isNaN(dob.getTime())) return false

  let age = today.getFullYear() - dob.getFullYear()
  const monthDelta = today.getMonth() - dob.getMonth()
  if (monthDelta < 0 || (monthDelta === 0 && today.getDate() < dob.getDate())) age -= 1
  return age >= 18
}
